#! /bin/sh -e

# DP: Description: Make libpthread load indirectly 
# DP: Related bugs: #312488
#		    http://lists.gnu.org/archive/html/bug-hurd/2005-05/msg00029.html
# DP: Patch Author: Roland McGrath 
# DP: Dpatch author: Michael Banck <mbanck@debian.org>
# DP: Upstream status: Not applied
# DP: Date: 08 Jun 2005

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p0 < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p0 < $0;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

Index: sysdeps/mach/hurd/i386/init-first.c
===================================================================
RCS file: /cvs/glibc/libc/sysdeps/mach/hurd/i386/init-first.c,v
retrieving revision 1.43
diff -B -b -p -u -r1.43 init-first.c
--- sysdeps/mach/hurd/i386/init-first.c 13 Sep 2004 00:42:46 -0000
    1.43
+++ sysdeps/mach/hurd/i386/init-first.c 3 May 2005 00:11:14 -0000
@@ -54,7 +55,7 @@ extern int __libc_argc attribute_hidden;
 extern char **__libc_argv attribute_hidden;
 extern char **_dl_argv;
 
-void *(*_cthread_init_routine) (void); /* Returns new SP to use.  */
+extern void *(*_cthread_init_routine) (void) __attribute__ ((weak));
 void (*_cthread_exit_routine) (int status) __attribute__ ((__noreturn__));
 
 /* Things that want to be run before _hurd_init or much anything else.
@@ -203,7 +204,7 @@ init (int *data)
      code as the return address, and the argument data immediately above
      that on the stack.  */
 
-  if (_cthread_init_routine)
+  if (&_cthread_init_routine && _cthread_init_routine)
     {
       /* Initialize cthreads, which will allocate us a new stack to run on.  */
       int *newsp = (*_cthread_init_routine) ();

